src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Diagnostics/RazorPullDiagnosticResponse.cs

```diff
@@ -1,8 +1,50 @@
 ﻿// Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the MIT license. See License.txt in the project root for license information.
 
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.LanguageServer.Protocol;
 using Microsoft.VisualStudio.LanguageServer.Protocol;
 
 namespace Microsoft.AspNetCore.Razor.LanguageServer.Diagnostics;
 
-internal record RazorPullDiagnosticResponse(VSInternalDiagnosticReport[] CSharpDiagnostics, VSInternalDiagnosticReport[] HtmlDiagnostics);
+internal record RazorPullDiagnosticResponse(VSInternalDiagnosticReport[] CSharpDiagnostics, VSInternalDiagnosticReport[] HtmlDiagnostics, VSInternalDiagnosticReport[]? CSharpAdditionalDiagnostics)
+{
+    public int Length
+        => CSharpDiagnostics.Length
+        + HtmlDiagnostics.Length
+        + (CSharpAdditionalDiagnostics?.Length ?? 0);
+
+    internal async Task AppendDiagnosticsAsync(RazorTranslateDiagnosticsService translateDiagnosticsService, VersionedDocumentContext documentContext, List<VSInternalDiagnosticReport> allDiagnostics, CancellationToken cancellationToken)
```

Comment:
I'm fine with extracting this to a method, but I'm not sure I like it being a method on a DTO record. This could also just be a gut reaction to a first time thing.

Reply:
I don't know either tbh. I was just going to make it a method on the class, but then I figured I'd test the waters and see how it feels. 

I might just move this back to being a method in the endpoint. This works in my head because it means that appending is always the same wherever this object is used (domain knowledge of itself I guess?) but in the end I feel like that's some weirdly purist view that doesn't work in our code base or elsewhere. If the object exists in a place where it _could_ have all the pieces to do something, and needs to take data from itself to do that, is it the right object for the job? (smartthinking.gif)

Reply:
Your "weirdly purist" view is not wrong, and in fact overlaps with things I had to do with co-hosting, where we need to move more of our "business logic" out of LSP handlers, and into services/helper methods. DTOs could be added to that list if we wanted, so this code is actually more future-proof for cohosting that just extracting the method and leaving it on the handler.

Reply:
Yea that's kind of my thinking. It's basically an extension method, but since it's in the same scope as where I would have written the extension method, I just... added it here